Issue/5697 add deposit business day delay notice

james-allan opened the pull request:
Fixes #5697 

#### Changes proposed in this Pull Request

This PR utilises the changes in `issue/5798-add-banner-notice` (PR #5814) to add a notice to the Deposits signalling that deposits may take 1-2 business days. 


<p align="center">
<img width="697" alt="Screenshot 2023-04-11 at 2 35 17 pm" src="https://user-images.githubusercontent.com/8490476/231057293-1918246c-4a22-4e6c-8401-3a65c12c73ec.png">
</p>

#### Testing instructions

1. Checkout this branch.
2. Enable the deposits overview updates by flipping on the feature flag.
   - `update_option( '_wcpay_feature_simplify_deposits_ui', 1 );`
3. Go to the **Payments > Overview** admin page.
4. Depending on your store you will either see one of the following results. 

| 1 Pending        | 1 In-transit           | All Sent | 
|:-------------:|:-------------:|:-----:|
| <img width="697" alt="Screenshot 2023-04-11 at 2 35 17 pm" src="https://user-images.githubusercontent.com/8490476/231057293-1918246c-4a22-4e6c-8401-3a65c12c73ec.png"> | <img width="701" alt="Screenshot 2023-04-11 at 2 36 27 pm" src="https://user-images.githubusercontent.com/8490476/231057447-c4eb41ec-7d3f-49ff-9273-d802c6afba02.png"> | <img width="699" alt="Screenshot 2023-04-11 at 2 37 42 pm" src="https://user-images.githubusercontent.com/8490476/231057553-6b9d4e91-70cf-40ad-81bf-e35f762414d7.png"> | 




| Multiple pending/in-transit        | Only 1 In-transit | No deposit |
|:-------------:|:-------------:|:-----:|
|  <img width="699" alt="Screenshot 2023-04-11 at 2 31 32 pm" src="https://user-images.githubusercontent.com/8490476/231056709-599b028e-5e7f-4cb2-9017-3635401ed580.png"> |  <img width="696" alt="Screenshot 2023-04-11 at 2 30 27 pm" src="https://user-images.githubusercontent.com/8490476/231056549-77b094d5-0ebd-4e9c-bf60-d4bcec933b99.png">| <img width="699" alt="Screenshot 2023-04-11 at 2 39 00 pm" src="https://user-images.githubusercontent.com/8490476/231057715-7ed3819b-75f9-4aae-83ab-188ca160471e.png">   |




To test the various statuses I had to mock the response from `useRecentDeposits`. For example, here's a code snippet of what I did locally to easily test the various deposit statuses without needing to wait for Stripe to generate a deposit. 

```php
const RecentDepositsList: React.FC< RecentDepositsProps > = ( {
	currency,
}: RecentDepositsProps ): JSX.Element => {
	//const recentDeposits = useRecentDeposits( currency );
	const recentDeposits = {
		isLoading: false,
		deposits: [
			{
				id: '1',
				date: '2023-04-10',
				type: 'deposit',
				amount: 2982,
				currency: 'USD',
				fee_percentage: 0,
				fee: 0,
				status: 'in_transit',
				bankAccount: '123456789',
				automatic: true,
			} as CachedDeposit,
			{
				id: '2',
				date: '2023-03-09',
				type: 'deposit',
				amount: 13721,
				currency: 'USD',
				fee_percentage: 0,
				fee: 0,
				status: 'paid',
				bankAccount: '987654321',
				automatic: true,
			} as CachedDeposit,
			{
				id: '2',
				date: '2023-02-09',
				type: 'deposit',
				amount: 12314,
				currency: 'USD',
				fee_percentage: 0,
				fee: 0,
				status: 'paid',
				bankAccount: '987654321',
				automatic: true,
			} as CachedDeposit,
		],
	};
	const isLoading = recentDeposits.isLoading;
...
```




-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `5839` or branch name `issue/5697-add-deposit-business-day-delay-notice` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=issue/5697-add-deposit-business-day-delay-notice) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 46cad35dbefdda7f8314faadef50950d743e7e5f
- Build time: 2023-04-12 00:32:40 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Jinksi commented on client/components/deposits-overview/style.scss:
```diff
@@ -18,6 +18,12 @@
+	&__notices {
+		padding: 0 24px;
+		.wcpay-banner-notice.components-notice {
+			margin: 0;
```
I'm concerned we will be overriding margins like this for every BannerNotice we add.

I think that the margin being applied to `.wcpay-overview .components-notice` is too general.

https://github.com/Automattic/woocommerce-payments/blob/e4ea2e82609df2a198250d00cf2fb6cec83d6ff9/client/overview/style.scss#L30-L33

I'm guessing the intention was for notices that are direct children of `.wcpay-overview` and we should fix it there (e.g. `.wcpay-overview > .components-notice`).

What do you think?

james-allan replied:
This `margin: 0` may still be necessary even if I do update the `.wcpay-overview > .components-notice`.

There is this margin styling added to all WP core `.components-notice`

<img width="323" alt="Screenshot 2023-03-28 at 11 12 53 am" src="https://user-images.githubusercontent.com/8490476/228101210-999c908e-77e4-4c42-9545-0261e3794d54.png">

And these new notices in particular need to be edge to edge of their container. 

Having said that, I agree that the style which applies to all notices on the wcpay overview is too general.

Jinksi replied:
> There is this margin styling added to all WP core .components-notice
Ah that's a bummer, apologies I didn't look deeper. I guess we'll just have to reset the margin each time we need to 👍

james-allan replied:
Yep. 

I've updated the `.wcpay-overview .components-notice` selector in [43d0b19](https://github.com/Automattic/woocommerce-payments/pull/5839/commits/43d0b19dbc0792dd5cd892aa5bfd862fd0b34a98) so it is more targeted. 

I'm wondering why this `.wcpay-overview .components-notice` was added.

This test mode notice has its own version of the `margin: 24px 0;` style and I can't see any other default notice on this screen. There are others I just don't know how to show them.
 
<img width="713" alt="Screenshot 2023-03-28 at 11 48 12 am" src="https://user-images.githubusercontent.com/8490476/228105724-b46ab544-de1e-4c57-8bf1-064bf43f2ca5.png">

In any case, the other notices appear on the top level (not nested in a Card) and so will meet the updated selector. I think you're right though, it seems odd to apply such a broad selector to all notices on this page.

Jinksi commented on client/components/deposits-overview/next-deposit.tsx:
```diff
@@ -110,6 +112,17 @@ const NextDepositDetails: React.FC< NextDepositProps > = ( {
+			{ /* Next Deposit Notice */ }
+			{ ! isLoading && (
+				<div className="wcpay-deposits-overview__notices">
+					<BannerNotice
```
Do you think we should centre the vertical alignment of the `BannerNotice` icon & text with `align-items: center`?

`align-items: flex-start` (current)
![align-items-flex-start](https://user-images.githubusercontent.com/3147296/228079463-78655b9a-e983-4874-a802-e3cc52a3d80c.png)

`align-items: center`
![align-items-center](https://user-images.githubusercontent.com/3147296/228079459-bc0d07bd-68cd-4ef6-a4d7-78b3853a07d7.png)


This might be a change to `BannerNotice` rather than this parent component's CSS.

james-allan replied:
Yep this would be my preference too but it was inconsistent with the core notice designs. Because the icon needs to remain at the top. I couldn't do an `align-items:center` otherwise it would look like this:

<img width="441" alt="Screenshot 2023-03-28 at 10 26 27 am" src="https://user-images.githubusercontent.com/8490476/228095879-5614bac5-ab0c-47bc-abbb-cd71d3844fdb.png">

Thanks to this guide https://css-tricks.com/snippets/css/a-guide-to-flexbox/ that's proving to be super helpful, I think I can get around that with a `align-self: flex-start` on the icon. 

<img width="433" alt="Screenshot 2023-03-28 at 10 30 35 am" src="https://user-images.githubusercontent.com/8490476/228096337-1145d887-99a2-4180-9d6f-9cf82ebc1dba.png">

james-allan replied:
I'm wondering if I should apply these styles to the base element. I should right? This isn't a unique to this scenario.

james-allan replied:
I've updated the base banner notice styles in [c0dc054](https://github.com/Automattic/woocommerce-payments/pull/5839/commits/c0dc05459c8c0527b227547c07551064d0263c87).

james-allan commented:
**Update**: The designs for this notice have changed. This notice will now be part of the deposit history table. 

Given that the history table is still a WIP and there's a draft PR (https://github.com/Automattic/woocommerce-payments/pull/5850), this PR is now 🔴 **On hold** until the table is merged.

james-allan committed:
Add a banner notice to pending & in-transit deposits indicating delay

james-allan committed:
Remove testing code

james-allan committed:
Merge branch 'develop' into issue/5697-add-deposit-business-day-delay-notice

james-allan commented:
I've updated this branch to reflect the new designs as listed here: https://github.com/Automattic/woocommerce-payments/issues/5697#issuecomment-1486140287.

I've also updated the PR description with new screenshots and steps to test.

james-allan committed:
Add changelog entry exception

Jinksi commented:
Should we add some margin-bottom to the banner when there is a single `in_transit` deposit (perhaps if it is a last-child)?

![image](https://user-images.githubusercontent.com/3147296/231051165-6e5480ca-fba9-4dad-8dc5-20e08586bc66.png)

Jinksi commented on client/components/deposits-overview/style.scss:
```diff
@@ -41,7 +40,9 @@
 				}
 			}
```
I think this will do the trick for adding margin-bottom to a single deposit:
```suggestion
			&:last-child {
				margin-bottom: 16px;
			}
		}
```

Jinksi commented on client/components/deposits-overview/style.scss:
```diff
@@ -41,7 +40,9 @@
-
+		.wcpay-banner-notice.components-notice {
+			margin: 0 24px;
+		}
```
I think this will do the trick for [adding margin-bottom to a single deposit](https://github.com/Automattic/woocommerce-payments/pull/5839#issuecomment-1502647809):
```suggestion
			&:last-child {
				margin-bottom: 16px;
			}
		}
```

<img width="713" alt="image" src="https://user-images.githubusercontent.com/3147296/231052037-44b5872a-bfd3-4bda-af15-c313a8161b7c.png">

james-allan replied:
Oh good catch. Thanks!

james-allan replied:
fixed in [fe6236e](https://github.com/Automattic/woocommerce-payments/pull/5839/commits/fe6236e3e772d350cafbc70728bde5d37ab7882a)

Jinksi commented on client/components/deposits-overview/recent-deposits-list.tsx:
```diff
@@ -5,6 +5,7 @@ import * as React from 'react';
 import { Flex, FlexItem, Icon } from '@wordpress/components';
 import { calendar } from '@wordpress/icons';
 import { Link } from '@woocommerce/components';
+import HelpOutlineIcon from 'gridicons/dist/help-outline';
```
I may not be up to date with the discussion on the design, but should this icon be `notice-outline` rather than `help-outline`?

![image](https://user-images.githubusercontent.com/3147296/231053039-9d880efb-eb5d-4f74-8491-802be0a99e8f.png)

james-allan replied:
Yep I caught onto this and fixed it in 50359ddcf0205f28f90ebb2f22e7f25d5b7492ba. :)

james-allan committed:
Update the deposit history deposit delay notice to appear after the oldest deposit

Jinksi commented on client/components/deposits-overview/recent-deposits-list.tsx:
```diff
@@ -66,6 +68,39 @@ const RecentDepositsList: React.FC< RecentDepositsProps > = ( {
+			}
+
+			return (
+				<>
+					<DepositTableRow deposit={ deposit } />
+					{ bannerNotice }
+				</>
+			);
+		} )
```
Since this is returning an array of elements, we need to add a unique `key` prop. ([docs](https://react.dev/reference/react/Fragment#rendering-a-list-of-fragments)). You will see an error in the browser console: `Each child in a list should have a unique "key" prop.`.


```suggestion
				<Fragment key={ deposit.id }>
					<DepositTableRow deposit={ deposit } />
					{ bannerNotice }
				</Fragment>
			);
		} )
```

Note: this error existed before your changes, I've also fixed this in #5822.

james-allan replied:
Thanks :)

fixed in [491a673](https://github.com/Automattic/woocommerce-payments/pull/5839/commits/491a673cf27527672db45884099c7f63f68f2199).

james-allan committed:
Add margin to banner notice if its the last child

james-allan committed:
Add deposit.id as the key for elements generated as part of an array map

Jinksi commented on client/components/deposits-overview/recent-deposits-list.tsx:
```diff
@@ -66,6 +69,39 @@ const RecentDepositsList: React.FC< RecentDepositsProps > = ( {
+
+	// Add a notice indicating the potential business day delay for pending and in_transit deposits.
+	let bannerAdded = false;
+	const depositRows = recentDeposits.deposits
+		.reverse() // reverse the array so that the oldest pending or in_transit deposit has the notice added to it.
```
I don't see any immediate side-effect here, but since `reverse()` mutates the array which is defined outside of this component's scope, I suggest duplicating the array before mutating ([react docs](https://react.dev/learn/updating-arrays-in-state#updating-arrays-without-mutation)).
```suggestion
	const depositRows = [ ...recentDeposits.deposits ]
		.reverse() // reverse the array so that the oldest pending or in_transit deposit has the notice added to it.
```

james-allan replied:
Oh thanks. I was actually going to request specific feedback about this. I wasn't sure if there was a better way to achieve this.

Jinksi replied:
This _might_ be able to be improved by finding the ID of the last deposit _before_ executing the "render" loop, for example:

```tsx
	const lastPendingOrInTransitDepositId = [ ...recentDeposits.deposits ]
		.reverse()
		.find(
			( deposit ) =>
				'pending' === deposit.status || 'in_transit' === deposit.status
		)?.id;

	const depositRows = recentDeposits.deposits.map( ( deposit ) => (
		<Fragment key={ deposit.id }>
			<DepositTableRow deposit={ deposit } />
			{ deposit.id === lastPendingOrInTransitDepositId && (
				<BannerNotice
					status="info"
					icon={ <InfoOutlineIcon /> }
					children={ strings.notices.businessDayDelay }
					isDismissible={ false }
				/>
			) }
		</Fragment>
	) );
```

james-allan committed:
Update snapshots

james-allan committed:
Update client/components/deposits-overview/recent-deposits-list.tsx

Co-authored-by: Eric Jinks <3147296+Jinksi@users.noreply.github.com>

james-allan committed:
Merge branch 'develop' into issue/5697-add-deposit-business-day-delay-notice

james-allan commented on client/components/deposits-overview/recent-deposits-list.tsx:
```diff
@@ -108,12 +143,7 @@ const RecentDepositsList: React.FC< RecentDepositsProps > = ( {
-				{ deposits.map( ( deposit ) => (
-					<DepositTableRow key={ deposit.id } deposit={ deposit } />
-				) ) }
+				{ isLoading ? <DepositTableRowLoading /> : depositRows }
```
@Jinksi Let me know if I've interpreted this merge conflict correctly. 

It doesn't look like I have to be honest but keeping this logic the way it was, I was getting odd behavior of the loading elements showing for a split second whilst I had generated the rows. ie I had deposit data before isLoading was false. I don't know if that's possible in real environments.

Jinksi replied:
That looks like an improvement to the loading state logic to me and tests well 👍

Jinksi committed:
Merge branch 'develop' into issue/5697-add-deposit-business-day-delay-notice

Jinksi commented on client/components/deposits-overview/recent-deposits-list.tsx:
```diff
@@ -5,6 +5,8 @@ import * as React from 'react';
 import { calendar } from '@wordpress/icons';
 import { Link } from '@woocommerce/components';
+import InfoOutlineIcon from 'gridicons/dist/info-outline';
+import { Fragment } from '@wordpress/element';
```
TIL about `@wordpress/element`. I would have defaulted to importing `Fragment` from React, but I see that both are used throughout WCPay and I'm unaware of the guidelines.

haszari replied:
I'd recommend importing from `@wordpress` as the most high-level, appropriate API – we're using it as a wrapper/proxy for React in WordPress so should default to there first.

Jinksi commented on client/components/deposits-overview/recent-deposits-list.tsx:
```diff
@@ -93,6 +96,38 @@ const RecentDepositsList: React.FC< RecentDepositsProps > = ( {
 	}
 
+	// Add a notice indicating the potential business day delay for pending and in_transit deposits.
+	let bannerAdded = false;
+	const depositRows = [ ...deposits ]
+		.reverse() // reverse the array so that the oldest pending or in_transit deposit has the notice added to it.
+		.map( ( deposit ) => {
+			let bannerNotice = null;
+
+			if (
+				! bannerAdded &&
+				( 'pending' === deposit.status ||
+					'in_transit' === deposit.status )
+			) {
+				bannerNotice = (
+					<BannerNotice
+						status="info"
+						icon={ <InfoOutlineIcon /> }
+						children={ strings.notices.businessDayDelay }
+						isDismissible={ false }
+					/>
+				);
+				bannerAdded = true;
+			}
+
+			return (
+				<Fragment key={ deposit.id }>
+					<DepositTableRow deposit={ deposit } />
+					{ bannerNotice }
+				</Fragment>
+			);
+		} )
+		.reverse(); // reverse the array back to the original order.
```
To simplify this code, I'd suggest we determine the id of the deposit that needs the banner and provide it to the loop. Just my opinion, feel free to leave this as is, the outcome will be the same.

```suggestion
	// Add a notice indicating the potential business day delay for pending and in_transit deposits.
	const earliestPendingDepositId = [ ...deposits ]
		.reverse()
		.find(
			( deposit ) =>
				'pending' === deposit.status || 'in_transit' === deposit.status
		)?.id;

	const depositRows = deposits.map( ( deposit ) => (
		<Fragment key={ deposit.id }>
			<DepositTableRow deposit={ deposit } />
			{ deposit.id === earliestPendingDepositId && (
				<BannerNotice
					status="info"
					icon={ <InfoOutlineIcon /> }
					children={ strings.notices.businessDayDelay }
					isDismissible={ false }
				/>
			) }
		</Fragment>
	) );
```

james-allan replied:
Love it. This is much cleaner. Applied in [46cad35](https://github.com/Automattic/woocommerce-payments/pull/5839/commits/46cad35dbefdda7f8314faadef50950d743e7e5f).

Jinksi reviewed (approved):
This tests well as per instructions, nice one @james-allan 🙌 

I've left some minor suggestions that do not impact functionality, feel free to take or leave.

james-allan committed:
Refactor logic to deteremine the deposit that needs the business day notice applied
